cli/src/main/java/org/jboss/as/cli/CommandCompleter.java

```diff
@@ -124,6 +127,31 @@ protected int doComplete(CommandContext ctx, String buffer, int cursor, List<Str
             }
         }
 
+        // https://issues.jboss.org/browse/WFCORE-1714
+        if (unresolvedVariableException == false) {
+            int index = buffer.lastIndexOf('$');
+            if (index != -1) {
+                String variable = buffer.substring(index + 1);
+                if (buffer.endsWith(variable)) {
+                    List<String> candidateVariables = new ArrayList<>();
+                    for (String var : ctx.getVariables()) {
+                        if (var.startsWith(variable)) {
+                            candidateVariables.add(var);
+                        }
+                    }
+                    if (candidateVariables.size() > 1) {
+                        candidates.addAll(candidateVariables); // more than one candidate variables
+                        Collections.sort(candidates);
```

Comment:
I think that you shouldn't add candidates prior to properly complete the request. This is why you added the special case in OperationRequestCompleter.
You should add candidates after having called the completer.